File: Tests/DocShotTests/CaptureActivityPolicyTests.swift
import Testing
import Foundation
import CoreGraphics
@testable import DocShot

/// Only one DocShot capture activity at a time, decided without either coordinator knowing about
/// the other.
@Suite("CaptureActivityPolicy Tests")
struct CaptureActivityPolicyTests {

    private static let target = RecordingTarget.window(
        id: 1,
        boundsInCG: CGRect(x: 0, y: 0, width: 100, height: 100)
    )

    private static func recording() -> TemporaryRecording {
        TemporaryRecording(
            url: URL(fileURLWithPath: "/tmp/DocShot-Recordings/clip.mp4"),
            duration: 1,
            pixelSize: CGSize(width: 100, height: 100),
            hasAudio: false,
            createdAt: Date()
        )
    }

    @Test("With both subsystems idle, both capture actions are available")
    func testBothIdle() {
        let policy = CaptureActivityPolicy.evaluate(captureState: .idle, recordingState: .idle)

        #expect(policy.allowsScreenshotCapture)
        #expect(policy.allowsRecordingCapture)
        #expect(policy.allowsStopRecording == false)
    }

    @Test("Every non-idle recording state blocks screenshot capture")
    func testRecordingBlocksScreenshots() {
        let busyStates: [RecordingState] = [
            .permissionRequired,
            .selecting,
            .awaitingConfirmation(Self.target),
            .starting(Self.target),
            .recording(Self.target, startedAt: Date()),
            .stopping,
            .awaitingOutput(Self.recording()),
            .exportingGIF(Self.recording()),
            .failed(.streamStart("x"))
        ]

        for state in busyStates {
            let policy = CaptureActivityPolicy.evaluate(captureState: .idle, recordingState: state)
            #expect(policy.allowsScreenshotCapture == false, "Screenshot allowed during \(state)")
            #expect(policy.allowsRecordingCapture == false, "Second recording allowed during \(state)")
        }
    }

    @Test("An in-progress screenshot blocks a new recording")
    func testScreenshotBlocksRecording() {
        let busyStates: [CaptureState] = [.permissionRequired, .selecting, .capturing, .editing]

        for state in busyStates {
            let policy = CaptureActivityPolicy.evaluate(captureState: state, recordingState: .idle)
            #expect(policy.allowsRecordingCapture == false, "Recording allowed during \(state)")
        }
    }

    @Test("A cancelled screenshot leaves recording available")
    func testCancelledScreenshotAllowsRecording() {
        let policy = CaptureActivityPolicy.evaluate(captureState: .cancelled, recordingState: .idle)

        #expect(policy.allowsRecordingCapture)
    }

    @Test("Stop is offered only while a recording is opening or running")
    func testStopAvailability() {
        let stoppable: [RecordingState] = [
            .starting(Self.target),
            .recording(Self.target, startedAt: Date())
        ]
        for state in stoppable {
            #expect(CaptureActivityPolicy.evaluate(captureState: .idle, recordingState: state).allowsStopRecording)
        }

        let notStoppable: [RecordingState] = [
            .idle,
            .permissionRequired,
            .selecting,
            .awaitingConfirmation(Self.target),
            .stopping,
            .awaitingOutput(Self.recording()),
            .failed(.stopFailed("x"))
        ]
        for state in notStoppable {
            #expect(
                CaptureActivityPolicy.evaluate(captureState: .idle, recordingState: state).allowsStopRecording == false,
                "Stop offered during \(state)"
            )
        }
    }
}

File: windows/tests/DocShot.Core.Tests/CaptureActivityPolicyTests.cs
using DocShot.Core.Models;
using Xunit;

namespace DocShot.Core.Tests;

public class CaptureActivityPolicyTests
{
    [Fact]
    public void When_both_coordinators_are_idle_everything_is_allowed_except_stop()
    {
        var policy = CaptureActivityPolicy.Evaluate(CaptureState.Idle, RecordingState.IdleState);

        Assert.True(policy.AllowsScreenshotCapture);
        Assert.True(policy.AllowsRecordingCapture);
        Assert.False(policy.AllowsStopRecording);
    }

    [Fact]
    public void A_live_recording_blocks_screenshot_capture_from_starting()
    {
        var recording = new RecordingState.Recording(
            new RecordingTarget.Window(1, new DocShot.Core.Primitives.RectD(0, 0, 100, 100)),
            DateTimeOffset.UtcNow);

        var policy = CaptureActivityPolicy.Evaluate(CaptureState.Idle, recording);

        Assert.False(policy.AllowsScreenshotCapture);
        Assert.False(policy.AllowsRecordingCapture);
        Assert.True(policy.AllowsStopRecording);
    }

    [Fact]
    public void An_active_screenshot_capture_blocks_a_new_recording_from_starting()
    {
        var policy = CaptureActivityPolicy.Evaluate(CaptureState.Selecting, RecordingState.IdleState);

        Assert.False(policy.AllowsRecordingCapture);
        // Screenshot capture owns its own state guards; this policy does not additionally block it.
        Assert.True(policy.AllowsScreenshotCapture);
    }

    [Fact]
    public void Stop_is_only_allowed_while_starting_or_recording()
    {
        var target = new RecordingTarget.Window(1, new DocShot.Core.Primitives.RectD(0, 0, 100, 100));

        Assert.True(CaptureActivityPolicy.Evaluate(CaptureState.Idle, new RecordingState.Starting(target)).AllowsStopRecording);
        Assert.True(CaptureActivityPolicy.Evaluate(CaptureState.Idle, new RecordingState.Recording(target, DateTimeOffset.UtcNow)).AllowsStopRecording);
        Assert.False(CaptureActivityPolicy.Evaluate(CaptureState.Idle, new RecordingState.Stopping()).AllowsStopRecording);
        Assert.False(CaptureActivityPolicy.Evaluate(CaptureState.Idle, RecordingState.IdleState).AllowsStopRecording);
    }

    [Fact]
    public void Cancelled_capture_state_counts_as_idle_for_allowing_a_new_recording()
    {
        var policy = CaptureActivityPolicy.Evaluate(CaptureState.Cancelled, RecordingState.IdleState);
        Assert.True(policy.AllowsRecordingCapture);
    }
}
